### PR TITLE
Rust side of password reset

### PR DESCRIPTION
**Pull request by bitfriend**
Will fix [this issue](https://github.com/acterglobal/a3/issues/921).

### CODE REVIEW

**Comment by bitfriend** [comment]
During config of `mailhog`, I referenced [this document](https://stackoverflow.com/questions/60162842/how-can-i-add-basic-authentication-to-the-mailhog-service-in-ddev-local) and gist in that.

**Comment by bitfriend** [comment]
Fixed the following parse error in local docker compose by https://github.com/acterglobal/a3/pull/1783/commits/2c4a2865b6a3ff238aff00d9b928382c94d31e6d.
image
But this error doesn't appear in github actions workflow.

**Comment by bitfriend** [comment]
In email for password reset, mail sender is `synapse` and mail receiver is user.
And mail server is `mailhog` there.

**Comment by bitfriend** [comment]
Now I could confirm the following commands work fine.

- `sudo docker-compose up -d` at `./util/test_server` of project
- `cargo run -p acter-cli -- mock` at root of project
- `RUST_LOG="info" cargo test -p acter-test -- --nocapture msg_edit::edit_text_msg` at root of project

When generating mock data, `acter-cli` seems to access email handling via `register_with_token_under_config()`.
All of above commands can work fine, only when email config is perfect, once you started to configure email at `homeserver.yaml` via `ci-start.sh`.

**Comment by bitfriend** [comment]
Implemented `mailhog` in github actions workflow, based on [this document](https://blog.potamic.net/running-mailhog-in-github-actions).

**Comment by bitfriend** [comment]
Password reset under logged-in was already completed.
For password reset under not-logged-in, Olivier 'reivilibre' (@reivilibre.element:librepush.net) proposed the following procedure.

1. call `Account::change_password()` API, it returns a set of flows
2. look at the flows and see if you can perform any of them (for you: this means check that e-mail identity is accepted for a password reset)
3. for e-mail identity: request a 3pid token like you already do `Account::request_3pid_email_token()`
4. when the token is confirmed, you call the `/account/password` API again with the information and it should complete

I found this was correct guide, after my implementation was succeeded.

**Comment by bitfriend** [comment]
I changed logic ordering somewhat.

1. request email token using `Account::request_3pid_email_token()`.
2. read email msg and associate email address with account using `Account::add_3pid()`.
3. log out client.

User should associate email address with his account before he forgets his password.
Up to this point, it is assumed user is logged in.

4. call `request_password_change_token_via_email` API to start `m.login.email.identity` flow.
5. call `/_matrix/client/v3/account/password` API and extract `session` from its return.
6. read email msg from `request_password_change_token_via_email` API and extract token & submit URL from email msg.
7. if token submitted to URL, `m.login.email.identity` is completed.
8. use token/secret/sid from email msg & session from `/_matrix/client/v3/account/password` API & new password to call `/_matrix/client/v3/account/password` API again.
9. Now user can login with only new password not old password.

The steps of `#4` ~ `#9` are performed by `reqwest` with login.

File: native/test/src/tests/auth.rs [hunk 1]
```diff
@@ -288,7 +286,7 @@ async fn can_register_via_email() -> Result<()> {
         resp.submit_url(),
     );
 
-    read_email_msg("test2", "test", "_matrix/client/unstable/registration").await?;
+    confirm_email_msg(email, "_matrix/client/unstable/registration").await?;
```
Comment (bitfriend) [inline native/test/src/tests/auth.rs:289]: Ah, you did not only read email msg, but also submit there?
Great.

Comment (gnunicorn) [reply native/test/src/tests/auth.rs:290]: submitting was already there... the code on `read` would also do this before, it wasn't really clear though.

File: .github/workflows/integration_testing.yml [hunk 2]
```diff
@@ -73,6 +79,35 @@ jobs:
...
+                message_from_person_in_room: "[%(app)s] You have a message on %(app)s from %(person)s in the %(room)s room..."
+                message_from_person: "[%(app)s] You have a message on %(app)s from %(person)s..."
+                messages_from_person: "[%(app)s] You have messages on %(app)s from %(person)s..."
+                messages_in_room: "[%(app)s] You have messages on %(app)s in the %(room)s room..."
+                messages_in_room_and_others: "[%(app)s] You have messages on %(app)s in the %(room)s room and others..."
+                messages_from_person_and_others: "[%(app)s] You have messages on %(app)s from %(person)s and others..."
+                invite_from_person_to_room: "[%(app)s] %(person)s has invited you to join the %(room)s room on %(app)s..."
+                invite_from_person: "[%(app)s] %(person)s has invited you to chat on %(app)s..."
+                password_reset: "[%(server_name)s] Password reset"
+                email_validation: "[%(server_name)s] Validate your email"
```
Comment (gnunicorn) [inline .github/workflows/integration_testing.yml:109]: not sure these are necesary ... are we sure these are needed? I'd like to keep configuration changes in the tests at a minimum.

File: Cargo.toml [hunk 3]
```diff
@@ -8,63 +8,62 @@ members = [
 default-members = ["native/acter"]
 
 [workspace.dependencies.matrix-sdk]
-git = "https://github.com/matrix-org/matrix-rust-sdk"
-rev = "1380af4f005ee5f34607db5cbb6705f93048b85a"
+git = "https://github.com/acterglobal/matrix-rust-sdk"
+rev = "5e7a05c26f7b7029a3dc4176cf758c3cc459b04f"
```
Comment (gnunicorn) [inline Cargo.toml:12]: could you elaborate what changes had to be done to the matrix sdk to switch to a local fork? And what PR is pending to get those into upstream?

File: Cargo.toml [hunk 4]
```diff
@@ -8,63 +8,62 @@ members = [
...
-#version = "0.28.1"
+git = "https://github.com/acterglobal/ruma"
+rev = "c74a4dcc4ff6406288bbf83fa7475dc732fe3a5e"
 
 [workspace.dependencies.ruma-macros]
-git = "https://github.com/Hywan/ruma"
-branch = "feat-sliding-sync-list-include-heroes"
-#version = "0.13.0"
+git = "https://github.com/acterglobal/ruma"
+rev = "c74a4dcc4ff6406288bbf83fa7475dc732fe3a5e"
```
Comment (gnunicorn) [inline Cargo.toml:62]: what ruma changes are needed to have our own fork?

File: native/acter/api.rsh [hunk 5]
```diff
@@ -35,6 +35,15 @@ fn guest_client(base_path: string, media_cache_base_path: string, default_homese
...
 
+/// Request the registration token via email
+fn request_registration_token_via_email(base_path: string, media_cache_base_path: string, username: string, default_homeserver_name: string, default_homeserver_url: string, email: string) -> Future<Result<RegistrationTokenViaEmailResponse>>;
+
+/// Request the password change token via email
+fn request_password_change_email_token(default_homeserver_url: string, email: string) -> Future<Result<PasswordChangeEmailTokenResponse>>;
+
+/// Finish password reset without login
+fn reset_password(default_homeserver_url: string, sid: string, client_secret: string, new_val: string) -> Future<Result<bool>>;
+
```
Comment (gnunicorn) [inline native/acter/api.rsh:46]: I feel that the name of these is at least not super clear and the docs do not help explaining how to use them...

File: native/acter/src/api/account.rs [hunk 6]
```diff
@@ -152,6 +159,155 @@ impl Account {
             .await?
     }
 
+    pub async fn request_3pid_email_token(
```
Comment (gnunicorn) [inline native/acter/src/api/account.rs:162]: All these 3pid functions seem unrelated. do we need those for anything?

**Review by gnunicorn — APPROVED** [review]
A few changes but then this is good to go.

File: .github/workflows/integration_testing.yml (hunk 2 above)
Comment (bitfriend) [reply .github/workflows/integration_testing.yml]: Fixed by https://github.com/acterglobal/a3/pull/1783/commits/2e18a453ad8f45c0894307f47f2fd03a655a7283.
Left only required items.

File: Cargo.toml (hunk 4 above)
Comment (bitfriend) [reply Cargo.toml:62]: In fact, I needed `EmailIdentity::new()` and implemented it [here](https://github.com/ruma/ruma/commit/a02316c36a2eac042cf67ea9cad9db7bf27ac380).
But it is not useless now.
Will restore something like main branch.

File: Cargo.toml (hunk 3 above)
Comment (bitfriend) [reply Cargo.toml:12]: In fact, I needed `EmailIdentity::new()` and implemented it [here](https://github.com/matrix-org/matrix-rust-sdk/commit/7a2d5c5e5c88d8713e0b1dd5447f828b2efdcdfc).
But it is not useless now.
Will restore something like main branch.

Comment (bitfriend) [reply Cargo.toml]: Fixed by https://github.com/acterglobal/a3/pull/1783/commits/3935bc962c9995df31b4549587d2d8d0b0fd6a1d

File: Cargo.toml (hunk 4 above)
Comment (bitfriend) [reply Cargo.toml]: Fixed by https://github.com/acterglobal/a3/pull/1783/commits/3935bc962c9995df31b4549587d2d8d0b0fd6a1d
